Modules/VMware.Skyline.InsightsApi/VMware.Skyline.InsightsApi.psm1

```diff
@@ -0,0 +1,397 @@
+    Created by:	Brian Wuchner
+    Date:		February 21, 2022
+    Blog:		www.enterpriseadmins.org
+    Twitter:		@bwuch
+    ===========================================================================
+    .SYNOPSIS
+    Use this function to post a query to the Skyline Insights API.
+    .DESCRIPTION
+    This function will allow you to query the Skyline Insights API.
+    Proper headers will be formatted and posted if a DefaultSkylineConnection is present.
+    This is primarily a helper function used by other functions included in the module.  
+    It is exported in the module manifest to be used for any custom queries.
+    .EXAMPLE
+    PS C:\> Invoke-SkylineInsightsApi -queryBody '{formatted-query-string-converted-to-json}'
+#>
+    param(
+        [Parameter(Mandatory=$true)][string]$queryBody
+    )
+    
+    if ( !$global:DefaultSkylineConnection ) {
+        Write-Error 'You are not currently connected to any servers. Please connect first using Connect-SkylineInsights.'
+        return;
+    }
+
+    write-debug "Querybody: $queryBody"
+    try {
+        $restCall = invoke-restmethod -method post -Uri $($global:DefaultSkylineConnection.SkylineAPI) -Headers @{Authorization = "Bearer $($global:DefaultSkylineConnection.ConnectionDetail.access_token)"} -body $queryBody -ContentType "application/json"
+        if ($restCall.errors) {
+            Write-Error $restCall.errors.Message
+        }
+        return $restCall
+    } catch {
+        $incomingError = $_
+        try {
+            # are nested try/catch blocks the powershell equilivent of vbscript On Error Resume Next?
+            $errorStatusAsJson = ($incomingError | ConvertFrom-Json).status
+            if ($errorStatusAsJson -eq '429 TOO_MANY_REQUESTS') {
+                write-debug 'Hit 429 / rate limite warning, will sleep for 1 second.'
+                start-sleep -Milliseconds 1005
+                Invoke-SkylineInsightsApi $queryBody
+            }
+        } catch {
+            # this was the error from trying to cast the incoming error to Json
+        }
+        if (!$errorStatusAsJson) { write-error $_ }
+    }
+}
+
+
+Function Get-SkylineFinding {
+<#
+    .NOTES
+    ===========================================================================
+    Created by:	Brian Wuchner
+    Date:		February 21, 2022
+    Blog:		www.enterpriseadmins.org
+    Twitter:		@bwuch
+    ===========================================================================
+    .SYNOPSIS
+    Use this function to query findings from the Skyline Insights API.
+    .DESCRIPTION
+    This function will allow you to query the Skyline Insights API for Findings.
+    As described in the documentation, the maximum limit per page is 200 records.  This function provides
+    an optional pagesize parameter to request smaller batches, but by default assumes 200 records.
+    .EXAMPLE
+    PS C:\> Get-SkylineFinding
+#>
+    [cmdletbinding()]
+    param(
+        [ValidateRange(1,200)][int]$pagesize=200
```

Comment:
I would add additional filters here like "Name" for example. Severity and findingImpact also seem like good candidates. Even if you cannot make server side filtering by those (can you?) a client side filtering will be useful. In next iteration an additional cmdlet that retreieves the categories and filter by category might also be a good addition.

Reply:
I've added optional server side filtering for findingId and products as well as client side filtering for severity for the next commit.  I did not include findingImpact as it seems there are multiple possibilities for this value and I couldn't find a comprehensive list to include in a validateset.  Severity only has 3 choices, critical, moderate, and trivial so it was easier to accommodate.